soc: nordic: add IronSide SE compatible UICR support

jonathannilsen opened the pull request:
Add support for generating UICR and associated artifacts in a format compatible with IronSide SE, to be used for nRF SoCs in the Haltium family.

The main feature added with this is the ability to configure certain global domain peripherals that are managed by the secure domain through setting UICR.PERIPHCONF. This register points at a blob of (register address, register value) pairs which are loaded into the peripherals by IronSide SE ahead of the application boot.

The added helper macros in uicr.h can be used to add register configurations to the PERIPHCONF. Entries added through these macros are then extracted by a script, post-processed and placed in a blob located at specific part of MRAM.

A default PERIPHCONF configuration has been added for the nrf54h20 soc to support the standard BLE use case.

jonathannilsen committed:
soc: nordic: add IronSide SE compatible UICR support

Add support for generating UICR and associated artifacts in a
format compatible with IronSide SE, to be used for nRF SoCs
in the Haltium family.

The main feature added with this is the ability to configure certain
global domain peripherals that are managed by the secure domain
through setting UICR.PERIPHCONF. This register points at a blob of
(register address, register value) pairs which are loaded
into the peripherals by IronSide SE ahead of the application boot.

The added helper macros in uicr.h can be used to add register
configurations to the PERIPHCONF. Entries added through these macros
are then extracted by a script, post-processed and placed in a blob
located at specific part of MRAM.

A default PERIPHCONF configuration has been added for the nrf54h20
soc to support the standard BLE use case (matching the configuration
in the soc devicetree).

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

nordicjm commented on soc/nordic/common/CMakeLists.txt:
```diff
@@ -3,6 +3,10 @@
 
 add_subdirectory_ifdef(CONFIG_RISCV_CORE_NORDIC_VPR vpr)
 
+if((CONFIG_NRF_PERIPHCONF_SECTION OR CONFIG_NRF_HALTIUM_GENERATE_UICR))
```
excess brackets (nit.)

jonathannilsen committed:
west: runners: nrf: Program UICR/PERIPHCONF artifacts if present

Program the new UICR and PERIPHCONF artifacts if they are generated.
These are required for the application to operate properly if they
are in use.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

karstenkoenig commented on soc/nordic/common/uicr/uicrgen.py:
```diff
@@ -0,0 +1,272 @@
+    return bytes(final_periphconf)
+
+
+def parse_kconfig(content: str) -> dict[str, str | None]:
```
Hmm is this the first time a .config file gets loaded as a dictionary and there is no utility library somewhere?

karstenkoenig commented on soc/nordic/common/uicr/uicrgen.py:
```diff
@@ -0,0 +1,272 @@
+    return result
+
+
+def format_hexdata(data, per_line: int = 16) -> str:
```
Is this used anywhere?

jonathannilsen replied:
No, it was used by a now deleted function, will delete it.

karstenkoenig reviewed (commented):
This looks great, the spu macros make the whole thing nicely readable

jonathannilsen commented on soc/nordic/common/uicr/uicr.h:
```diff
@@ -0,0 +1,288 @@
+	UICR_PERIPHCONF_ADD((uint32_t)&((NRF_SPU_Type *)(_spu))->FEATURE.GRTC.CC[_index],          \
+			    _UICR_SPU_FEATURE_VAL(_secattr, _ownerid))
+
+/* Common macro for encoding a SPU FEATURE.* register value. */
```
Added a comment

jonathannilsen commented:
Some changes here are now out of sync with the version of this in #91826, closing.
